[RISCV] Pack build_vectors into largest available element type

preames opened the pull request:
Our worst case build_vector lowering is a serial chain of vslide1down.vx operations which creates a serial dependency chain through a relatively high latency operation.  We can instead pack together elements into ELEN sized chunks, and move them from integer to scalar in a single operation.

This reduces the length of the serial chain on the vector side, and costs at most three scalar instructions per element.  This is a win for all cores when the sum of the latencies of the scalar instructions is less than the vslide1down.vx being replaced, and is particularly profitable for out-of-order cores which can overlap the scalar computation.

This patch is restricted to configurations with zba and zbb.  Without both, the zero extend might require two instructions which would bring the total scalar instructions per element to 4.  zba and zba are both present in the rva22u64 baseline which is looking to be quite common for hardware in practice; we could extend this to systems without bitmanip with a bit of extra effort.

preames committed:
[RISCV] Pack build_vectors into largest available element type

Our worst case build_vector lowering is a serial chain of vslide1down.vx
operations which creates a serial dependency chain through a relatively
high latency operation.  We can instead pack together elements into ELEN
sized chunks, and move them from integer to scalar in a single operation.

This reduces the length of the serial chain on the vector side, and costs
at most three scalar instructions per element.  This is a win for all
cores when the sum of the latencies of the scalar instructions is less
than the vslide1down.vx being replaced, and is particularly profitable
for out-of-order cores which can overlap the scalar computation.

This patch is restricted to configurations with zba and zbb.  Without
both, the zero extend might require two instructions which would bring
the total scalar instructions per element to 4.  zba and zba are both
present in the rva22u64 baseline which is looking to be quite common
for hardware in practice; we could extend this to systems without
bitmanip with a bit of extra effort.

preames commented:
Quick note for reviewers, please pay close attention to byte order issues in this.  I frequently get that wrong, and while this is inspired by a downstream patch I rewrote this basically from scratch.  Definitely room for error.

topperc commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll:
```diff
@@ -1267,43 +1267,53 @@ define <16 x i8> @buildvec_v16i8_loads_contigous(ptr %p) {
 ; RVA22U64:       # %bb.0:
-; RVA22U64-NEXT:    addi a6, a0, 8
-; RVA22U64-NEXT:    lbu t6, 1(a0)
+; RVA22U64-NEXT:    lbu a1, 1(a0)
```
Should we test RVA22U32?

preames replied:
Not sure it adds any useful coverage, but happy to add it if desired.

topperc replied:
I asked specifically because it allows ELEN=64 with XLen=32.

preames replied:
We don't appear to have support in tree for rva22u32 (as an mattr alias).  I went digging through profile docs, and I can't find any mention of the rva22u32 profile variants; rva22 appears to only define 64 bit versions.

Do you have a particular arch string for a configuration which exercises the case you're concerned about here?

preames replied:
From your other comment, it seems like plain rv32 + v + zba + zbb might be enough here?  Can you confirm?

topperc replied:
I hadn't realized that that the RVA* profiles are 64-bit only. rv32 + v + zba + zbb is enough.

preames commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -3896,6 +3896,66 @@ static SDValue lowerBuildVectorOfConstants(SDValue Op, SelectionDAG &DAG,
+  for (unsigned i = 0; i < VT.getVectorNumElements(); i += 2) {
+    SDValue A = Op.getOperand(i);
+    SDValue B = Op.getOperand(i + 1);
+    if (ElemVT != XLenVT) {
```
I think you're right, this condition is just dead.  Will remove.

preames commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -3896,6 +3896,66 @@ static SDValue lowerBuildVectorOfConstants(SDValue Op, SelectionDAG &DAG,
+
+  unsigned NumElts = VT.getVectorNumElements();
+  unsigned ElemSizeInBits = ElemVT.getSizeInBits();
+  if (ElemSizeInBits >= Subtarget.getELen() || NumElts % 2 != 0)
```
Do we allow configurations where ELEN > XLEN?  I couldn't find a clear answer in the specification as to whether that was a valid point in the configuration cross product.

topperc replied:
Zve64* on RV32 allows that

preames replied:
Added, and rebased.  Turned out to be a good thing, as the configuration exposed a crash (which is now fixed).

preames committed:
Address a review comment

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -3896,6 +3896,64 @@ static SDValue lowerBuildVectorOfConstants(SDValue Op, SelectionDAG &DAG,
+
+  unsigned NumElts = VT.getVectorNumElements();
+  unsigned ElemSizeInBits = ElemVT.getSizeInBits();
+  if (ElemSizeInBits >= Subtarget.getELen() || NumElts % 2 != 0)
```
Do we ever end up with a non power of 2 VT during lowering? If not we could move `NumElts % 2 != 0` to an assert

topperc replied:
1 element vectors?

preames replied:
I added tests for 1 element vectors, just to be sure we had coverage here.

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll:
```diff
@@ -2171,3 +2209,240 @@ define <16 x i8> @buildvec_v16i8_loads_undef_scattered(ptr %p) {
+  ret <4 x i16> %v4
+}
+
+define <2 x i32> @buildvec_v2i32_pack(ptr %p, i32 %e1, i32 %e2) {
```
ptr %p looks to be unused in these tests

lukel97 reviewed (approved):
LGTM with nits

preames committed:
Merge branch 'main' into pr-pack-buildvector

preames committed:
Address review comments

preames commented:
@topperc Have I addressed all of your concerns?  This has an LGTM, but you pointed out a substantial problem after that, so I don't want to land without confirmation.

preames committed:
Merge branch 'main' into pr-pack-buildvector

topperc reviewed (approved):
LGTM
